### PR TITLE
Add heartbeat capability

### PR DESCRIPTION
**Pull request by kpamnany**
## PR Description

Telemetry for Julia's threads.

This capability is build-time enabled by setting `JL_HEARTBEAT_THREAD` in `src/options.h`, which causes Julia to start a heartbeat thread that quickly goes to sleep.

Control heartbeats with `jl_heartbeat_enable(heartbeat_s, n_reports, reset_reporting_after_s)`. Heartbeats are disabled if `heartbeat_s <= 0`. Otherwise, you must call `jl_heartbeat()` at least one time every `heartbeat_s`. If this does not happen, the heartbeat thread will call `jl_print_task_backtraces(0)`. If `jl_heartbeat()` continues to not be called, we double the reporting interval up to a maximum of 60 times the `heartbeat_s` and report up to `n_reports` times altogether. Once `jl_heartbeat()` is called again, the number of reports is reset immediately but we only reset the reporting interval back to `heartbeat_s` after `reset_reporting_after_s` of steady `jl_heartbeat()`s are observed.

## Checklist

Requirements for merging:
- ~~[ ] I have opened an issue or PR upstream on JuliaLang/julia:~~
- [X] I have created an issue to upstream this feature: https://relationalai.atlassian.net/browse/RAI-17788
- [X] I have removed the `port-to-*` labels that don't apply.
- [X] I have opened a PR on raicode to test these changes: https://github.com/RelationalAI/raicode/pull/15844

### CODE REVIEW

File: src/threading.c
Comment (NHDaly) [inline src/threading.c:851]: This could probably go in its own .c file?

File: src/threading.c
Comment (NHDaly) [inline src/threading.c:946]: 🤔 I'm confused: I thought we explicitly did _not_ want the heartbeat thread to be a julia thread. What does adopt thread do here? This seems surprising to me.

If it's here for a good reason, which I assume it is, can you please add comments explaining what's going on here? Surprising code should be commented. :)

File: src/threading.c
Comment (NHDaly) [inline src/threading.c:949]: Ah, I see, but then you enter a GC Safe region so that the other threads never need to wait on this thread. Interesting!

ohhhhh i see, we need the ptls so that we can call jl_print_task_backtraces? Fascinating. Yeah, all of this needs to be commented more heavily, though i think i'm piecing it together.

File: src/threading.c
Comment (NHDaly) [inline src/threading.c:951]:
```suggestion
    while (true) {
```
IMO this is better style 🤷

File: src/threading.c
Comment (NHDaly) [inline src/threading.c:932]: IMO it's nice to have some kind of exponential backoff on the reporting interval for expensive reports like this. Something like `report_interval_s *= 1.5` or something.

Up to you though. If you set it large enough it's probably fine, but I think that reporting the task backtraces can be pretty slow for us in raicode, since stack unwinding + symbolizing is really slow in our huge binary, and I'm worried that this brownout reporting can itself cause some brownouts? 😰 So I would probably suggest backing off the reporting interval with each report?

And then maybe shortening the interval again if we get one and it's been like 100x the reporting interval or something?
I guess this can be kind of finicky, but it's the kind of trick I've seen @tveldhui use to reduce the impact of expensive logs like this.

**Review by NHDaly — COMMENTED** [review]
This mostly LGTM! I think it needs a couple tests?

Also, should we open this as a PR upstream? This seems like a useful feature for anyone who is experiencing process hangs and/or deadlocks, not just for us, yeah?

File: src/threading.c
Comment (kpamnany) [reply src/threading.c:851]: It could, yes. That would require touching `Makefile`s so I decided to keep the patch small for now.

File: src/threading.c
Comment (kpamnany) [reply src/threading.c:946]: I'll add more comments. You have the answer below.

File: src/threading.c
Comment (kpamnany) [reply src/threading.c:949]: When a backtrace is gathered, it is written to the thread's TLS. So this thread needs a TLS and hence we have to call `jl_adopt_thread` to get one.

File: src/threading.c
Comment (kpamnany) [reply src/threading.c:951]: Some C compilers used to (still do?) complain about constants in loop conditions. But not about empty `for` statements. 🤷‍♂️

Comment (NHDaly) [reply src/threading.c:951]: haha gross. Okay, cool, feel free to ignore this then.

**Comment by NHDaly** [comment]
My main remaining question is that this should probably also be opened upstream, right?
Also, what happened to the PR checkboxes? Can you add those back?

**Comment by NHDaly** [comment]
Makes sense. Please file an internal tracking issue then, so we don't forget, and put that in place of the first checkbox. And attend to the second checkbox.

Assuming those are addressed and you add a bunch of comments, this LGTM i think!

**Comment by kpamnany** [comment]
More comments added. Issue to upstream this created and linked.

Please review the sleep duration computation I've changed in [these lines](https://github.com/RelationalAI/julia/pull/94/files#diff-d9f90322339ad42828d9a13cf46f8e75606098a0b66a0e3a26ad2c0b1e3905d2R983-R991)? Previously, the code unsafely assumed that `tchb < 1e9`. This assumption is now removed.

**Comment by NHDaly** [comment]
> I spent some time trying to figure out how to test this. The challenge is that heartbeats are time driven and all timing sensitive tests are flaky in CI. Would love any ideas.

The standard way to deal with time-sensitive approaches is to mock out the timing element, and step the system forward in the tests.

There are basically two or three main variants that I've seen on that:
- mockable / test-injectable timing system
    - At google, any class that wanted to anything based on time would depend on some Timing or Clock interface, and so to instantiate one, you'd have pass in a `Time` object. In production, this would be a normal Time implementation, but in tests, you could use a `MockTime` instance, which lets you control the time from your tests.
- Insert mockable "pause points" specifically for testing, and have the tests override those mocks.
    - That's what we do in RAI. You can look at the v2-api tests for an example of this, where we have mockable pause points in the transaction processing, and we use a channel to set up synchronization between the client and server through those mocked pause points.
- (variant on the above) If your system already has some synchronization interaction, mock out those to drive the tests.
    - So if the system waits for a client before proceeding at some points, you can mock out the client and control the system that way.

I'm not sure which of those would be most applicable in this situation. Notably, Julia's codebase does not employ any of these in its current test suites, so the infrastructure isn't there to easily build upon it.

**Comment by NHDaly** [comment]
One way to mock out the timing would be: instead of calling `jl_hrtime()` directly, you could call it through a global function pointer which _defaults to_ jl_hrtime. Then the tests could update that global function pointer to point to a mocked timing function, which calls a julia callback instead, and from there you can now do whatever you want to step time forward.

File: src/threading.c
Comment (kpamnany) [reply src/threading.c]: Added a 1.5X backoff to the reporting interval up to a maximum of 50X the requested reporting interval.

**Comment by kpamnany** [comment]
As discussed, I've changed the configuration and behavior of the heartbeat thread; see [here](https://github.com/RelationalAI/julia/pull/94/files#diff-d9f90322339ad42828d9a13cf46f8e75606098a0b66a0e3a26ad2c0b1e3905d2R888-R892).

Additionally, I've hardened the interface so that two consecutive `jl_heartbeat_enable()` calls don't break things.

Here's the [log](https://hydra.dev.relational.ai/log/jlbh84xi1b4amp6f2plkfkyff9p3pnrw-raicode.drv) of the `benchmark.micro.query_evaluation` job with the heartbeat thread enabled; there are quite a few heartbeat losses -- task backtraces printed. There seem to be only two types:
- `do_bench` runs on thread 1 and occupies it completely; an artifact of how the benchmark is written. But it isn't clear why the second interactive thread doesn't pick up and run the server metrics task; it seems to be asleep.
- A thread is compiling (deep in LLVM) and numerous other threads are blocked on the `codegen_lock`.

Would love to try and get this complete and merged tomorrow.

**Comment by NHDaly** [comment]
Really neat to hear you're seeing this in action!

> * But it isn't clear why the second interactive thread doesn't pick up and run the server metrics task; it seems to be asleep

Is this because thread 1 isn't able to drive the event loop while it's running the do_bench function?

**Review by NHDaly — APPROVED** [review]
LGTM, thanks Kiran! 🎉 
Can you make sure all the open threads are resolved before merging? Otherwise LGTM

File: src/threading.c
Comment (kpamnany) [reply src/threading.c:851]: Leaving a note to resolve this, we should consider reorganizing the code when we upstream this.

File: src/threading.c
Comment (kpamnany) [reply src/threading.c]: Changed this to a 2X backoff in reporting time.

**Comment by Drvi** [comment]
I've ported the PR to all +RAI branches

**Pull request merged**
